### ISSUE TITLE
[BUG|Documentation error ] Possible bug with WpTestCase and Transactions.

### ISSUE THREAD

**Issue by calvinalkan** [lucatume/wp-browser#494]
In the docs for the WpLoader Module is says that: 

"Inspection of the database during tests, e.g. stopping execution using XDebug, will not show any change in the database. Keep this in mind while trying to debug integration tests using the WPLoader module.
**When configured to only load WordPress (loadOnly: true) then any database operation will be committed and written to the database.**" 

However, this is not the case when you Test extends the WpTestCase class. 

Inside  the  ``_setUp()`` method of WpTestCase ``start_transaction()`` is always called. The same goes for ``_tearDown``. 

I'm not sure if this is the expected behaviour if it's a bug. Reading the docs on using WpLoader with WpDB I would say it's a bug.  I m using the WpDb module but still want to use most of the WpTestCase functionality.

IMO there should be a conditional check like there are in the WpLoader Class. 

```php

// in _setUp()

 $loadOnly = ! empty( // get from config );
 
 if ( ! $loadOnly ) {
 
 $this->start_transaction. 
 
 }
 
```

The same thing for ``_tearDown``. 

Again maybe I'm understanding this badly. But reading the docs it looks like loadOnly: true should not use Transactions in Order to be able to debug using Xdebug. 

Another thing that is not completely clear to me: When using loadOnly: true, inside the WpLoader class, whats the difference between this `bootstrapWp` and `installAndBootstrapInstallation()`

```php 

if (! empty($this->config['loadOnly'])) {
            $this->bootstrapWP();
        } else {
            $this->installAndBootstrapInstallation();
        }

```

Thank you , 

Calvin

**Comment by lucatume** [comment]
Hi @calvinalkan, thanks for opening the issue.

Ideally, the `WPLoader` module should be used in `loadOnly: true` mode *only in acceptance and functional tests*, where the `WPdb` module will handle and set up the database by means of a dump, where one is either writing [Cest or Cept kind of tests](https://codeception.com/docs/07-AdvancedUsage#Cest-Classes), *not* when writing integration tests extending the `WPTestCase` class.
The correct behavior should be the code to throw an exception when using a `WPTestCase` and `WPLoader` is configured to use `loadOnly: true` - they should simply not be mixed and most users are simply confused by it.
I've not closed that "loop-hole" as there are some cases where _some_ users are leveraging that will full understanding of what is happening under the hood: set up the DB using the `WPDb` module and run transactional queries during tests in integration tests.

The difference between `bootstrapWP` and `installAndBootstrapInstallation` is that the first will only load WordPress in the current test context, without installing it and activating the required theme and plugin, the latter will install WordPress, activate theme and plugins, and then run the tests in a clean and ready WordPress installation.
The latter is how the PHPUnit-based Core testing suite works.

In short: the documentation is aimed at the majority of the users that will use `WPLoader` with `loadOnly: false` or not set in integration ("WordPress unit") tests and will use it with `loadOnly: true` in acceptance and functional tests. From past experiences, too much information is not helping, but I'm open to suggestions as to improve it.

What is your current setup? I'm curious. What are you trying to do? It would help me understand your current setup and frame the conversation.

**Comment by calvinalkan** [comment]
I see, I have been using this combination the whole time but there never once was an exception thrown. But yea its kind of confusing with WpTestCase and WpLoader. 

My use case is the following: 

I'm currently developing an open-source project myself that aims to replicate laravels eloquent orm in a lean and WordPress-optimized way. 

Eloquent is used as a query builder but `$wpdb` serves as the connection instead of opening a  new one. 

Thus you can imagine that my tests are highly coupled to `$wpdb`. What I have been doing so far is using the WpDB module because all the WpDb helper functions are essential to test my own code. But I also need to correctly reset the global state after each test which did only work reliably extending `WpTestCase`. But then I was not able to debug anything in sequel ace during tests which I'm sure you can imagine is essential for this use case. 

I'm starting to doubt whether my approach is the correct one with this. Tests run WAY too slow for obvious reasons. But in general, I very much like to follow the classical TDD approach where mocking is kept to a minimum. 

Now, in my code `wpdb::prepare` for the most part, is my entry point to the `wpdb class` but mocking `wpdb` in my tests seems very risky. It is over 3500+ overall and you never quite know how it reacts to certain input. So there would have to be made a lot of assumptions. 

The official eloquent package is integration tests using an in-memory SQL lite database which obviously is not an option here. I was wondering whether you may have another idea of how you would design this testsuite. 

Greetings, 

Calvin

**Comment by lucatume** [comment]
Thanks for taking the time to provide context and an explanation.

First: there is no *wrong* way of testing as long as you get the job done. I'm not saying this for good feeling: I think "right and wrong ways of testing" *really* hurt testing adoption and put in place useless gate-keeping that confuses adopters when they start, and practitioners when they face each other. Providing help and guidance is one thing, deeming someone work "right" or "wrong" is another matter entirely without a good understanding of the context.

So: I'm glad you set up to test your plugin, good, work, keep it on.

Now that the "purity" argument is out of the way, I have some ideas:

1. Detecting, in the context of a `WPTestCase` if the `loadOnly` configuration parameter is `true` or `false` should not be difficult, acting on that information accordingly should not be difficult either. I could explore how to detect that and see what disaster will arise, it might be a good improvment.
2. Slow... If `WPDb.cleanup` is set to `true`, then each test will see the database wiped; it _should not_ happen when using the `WPTestCase`, though, so I could use a trimmed down version of your testing configuration and an example test case.
3. I agree with you about mocking the `wpdb` class: unless it's a one-method job, it would only make your testing not that robust and you would end up having to mock too much.
4. The `WPTestCase` class, when used in conjunction with the `WPLoader` module only, no WPDb and `loadOnly` not set or set to `false` does a pretty good job at resetting the db state (as it uses transactions) between tests and will do that pretty fast. I would use that. If I'm guessing correctly, you're using the `WPDb` module to load a dump that contains a know start state of the DB you need to act on in your tests: I would remove that and set up the fixture you need on a tess-per-test basis using the `factory()` that comes with `WPLoader` (from the PHPUnit based Core Test Suite).

Again: I'm guessing here, but say you need to test fetching posts and you need the db to contain 100 posts before your test starts:

```php
public function test_offset(){
	$drafts = static::factory()->post->create_many(50, ['post_status' => 'draft']);
	$sorted_drafts = wp_list_sort(array_map('get_post', $drafts), 'post_date', 'ASC');
	$published = static::factory()->post->create_many(50, ['post_status' => 'publish']);
	
	$builder = new Query_Builder();
	// Suppose we get post IDs.
	$fetched = $builder->where('post_status', 'draft')
		->offset(10)->limit(7)->fetch();
	
	$this->assertEquals(wp_list_pluck(array_slice($sorted_drafts, 10, 7), 'ID'), $fetched)
}
```

This is just sudo code, especially in the Builder part, but the concept is to move the fixture set up from the database dump to each test: usually the time it takes to set up the fixture using factories is way faster than the one it takes to do it use WPDb. 

Keep using WPDb, with no dump specified, to have it just provide database assertion methods.

This is sudo-code, but it could provide you with some ideas.

Let me know.

**Comment by calvinalkan** [comment]
Thank you for your suggestions. 

Yea I already figured that I could use WpDb just for the tests methods. I did that and it drastically speeds up integration tests for sure. 

But it brings its different set off issues. 

For example, I currently have a working implementation of the laravel schema builder done in WordPress using wpdb. 

In other tests ( models, query builder etc), I'll use the schema builder to set up test tables and tear them down after. This is way faster than the wpdb dump. **BUT**, testing the schema builder itself is not really possible with the `WpTestCase`. 

Since many of the methods are DDL queries that auto-commit any transaction, thus bypassing the `WpTestCase` transactions. Not really a problem with wp-browser tho. I could however use a different suite for these kind of tests where I use the wpdb module to restore state. 


Regarding 3). 

I went for using a symlinked db.php database drop-in that holds `wpdb` object as a property and via a factory class receives the underlying `mysqli` property of the `wpdb` class. This way I can use real prepared statements not the shitty string escaping of `wpdb`. As a benefit, in other parts of the module I can just mock out my own db-drop-in and don't have to deal with mocking `wpdb`.

**Comment by lucatume** [comment]
@calvinalkan reading this, it looks like you've found a way to get to the testing setup you're looking for, am I correct?

If so, would you mind letting me know if I can close the issue or if there's some other issue you ran into?

**Comment by calvinalkan** [comment]
Yes, I'm closing this. Thank you very much for the help.

**Comment by lucatume** [comment]
I wanted to react with a thumbs-up. Then clicked the wrong icon: thumbs-up and thanks.
